pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go

```diff
@@ -0,0 +1,191 @@
+
+		return err
+	}
+
+	if !equality.Semantic.DeepEqual(cr.Rules, gr.InheritedFleetWorkspacePermissions.ResourceRules) {
+		// undo modifications if cr has changed
+		cr.Rules = gr.InheritedFleetWorkspacePermissions.ResourceRules
+		_, err := h.crClient.Update(cr)
+
+		return err
+	}
+
+	return err
+}
+
+func (h *fleetWorkspaceRoleHandler) reconcileWorkspaceVerbs(gr *v3.GlobalRole) error {
+	crName := wrangler.SafeConcatName(gr.Name, fleetWorkspaceVerbsName)
+	cr, err := h.crCache.Get(crName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	crMissing := apierrors.IsNotFound(err)
+
+	if gr.InheritedFleetWorkspacePermissions.WorkspaceVerbs == nil {
+		return h.deleteClusterRoleIfNeeded(!apierrors.IsNotFound(err), crName)
+	}
+
+	workspacesNames, err := h.fleetWorkspaceNames()
+	if err != nil {
+		return err
+	}
+	if len(workspacesNames) == 0 {
+		// skip if there are no workspaces besides local
+		return nil
+	}
+	desiredCR := backingWorkspaceVerbsClusterRole(gr, crName, workspacesNames)
+	if crMissing {
+		_, err := h.crClient.Create(desiredCR)
+
+		return err
+	} else if !equality.Semantic.DeepEqual(cr.Rules, desiredCR.Rules) {
+		// undo modifications if cr has changed
+		cr.Rules = desiredCR.Rules
+		_, err := h.crClient.Update(cr)
+
+		return err
+	}
+
+	return nil
+}
+
+func backingResourceRulesClusterRole(gr *v3.GlobalRole, crName string) *v1.ClusterRole {
+	return &v1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: v3.GlobalRoleGroupVersionKind.GroupVersion().String(),
+					Kind:       v3.GlobalRoleGroupVersionKind.Kind,
+					Name:       gr.Name,
+					UID:        gr.UID,
+				},
+			},
+			Labels: map[string]string{
+				grOwnerLabel:                wrangler.SafeConcatName(gr.Name),
+				controllers.K8sManagedByKey: controllers.ManagerValue,
+			},
+		},
+		Rules: gr.InheritedFleetWorkspacePermissions.ResourceRules,
+	}
+}
+
+func backingWorkspaceVerbsClusterRole(gr *v3.GlobalRole, crName string, workspaceNames []string) *v1.ClusterRole {
+	return &v1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: v3.GlobalRoleGroupVersionKind.GroupVersion().String(),
+					Kind:       v3.GlobalRoleGroupVersionKind.Kind,
+					Name:       gr.Name,
+					UID:        gr.UID,
+				},
+			},
+			Labels: map[string]string{
+				grOwnerLabel:                wrangler.SafeConcatName(gr.Name),
+				controllers.K8sManagedByKey: controllers.ManagerValue,
+			},
+		},
+		Rules: []v1.PolicyRule{
+			{
+				Verbs:         gr.InheritedFleetWorkspacePermissions.WorkspaceVerbs,
+				APIGroups:     []string{"management.cattle.io"},
+				Resources:     []string{"fleetworkspaces"},
+				ResourceNames: workspaceNames,
+			},
+		},
+	}
+}
+
+func (h *fleetWorkspaceRoleHandler) deleteClusterRoleIfNeeded(crExists bool, crName string) error {
```

Comment:
Nit: I don't think that we get a lot of value out of this side function. The owning function can determine when is appropriate to delete, and since the act of deleting is a one-line call, I don't think is saves us much.

Reply:
removed